GraalVM Native Image is broken on Mac

When I attempt to build an app with `native-image` installed from Nix store I'm getting

```
[info] [1/8] Initializing...                                                                                    (0.0s @ 0.24GB)
Error:  Error: Unable to detect supported DARWIN native software development toolchain.
Error:  Querying with command '/nix/store/5dbia7q27b3j96pfjv0fnkglc7457g35-clang-wrapper-16.0.6/bin/cc -v' prints:
Error:    clang version 16.0.6
Error:    Target: x86_64-apple-darwin22.6.0
Error:    Thread model: posix
Error:    InstalledDir: /nix/store/6ccjdf3miyls5k7vb192nr4ibyb6wvxl-clang-16.0.6/bin
Error:  Error: To prevent native-toolchain checking provide command-line option -H:-CheckToolchain
Error:  com.oracle.svm.core.util.UserError$UserException: Unable to detect supported DARWIN native software development toolchain.
Error:  Querying with command '/nix/store/5dbia7q27b3j96pfjv0fnkglc7457g35-clang-wrapper-16.0.6/bin/cc -v' prints:
Error:    clang version 16.0.6
Error:    Target: x86_64-apple-darwin22.6.0
Error:    Thread model: posix
Error:    InstalledDir: /nix/store/6ccjdf3miyls5k7vb192nr4ibyb6wvxl-clang-16.0.6/bin
Error:  To prevent native-toolchain checking provide command-line option -H:-CheckToolchain
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:138)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.codegen.CCompilerInvoker.addSkipCheckingInfo(CCompilerInvoker.java:107)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.codegen.CCompilerInvoker.<init>(CCompilerInvoker.java:75)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.codegen.CCompilerInvoker$DarwinCCompilerInvoker.<init>(CCompilerInvoker.java:280)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.codegen.CCompilerInvoker.create(CCompilerInvoker.java:85)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:973)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:569)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:537)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:526)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:701)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.start(NativeImageGeneratorRunner.java:140)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:95)
```

If I disable the toolchain check the error would be

```
 Error:  Error: Error compiling query code (in /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c). Compiler command '/nix/store/c1ib07kj7baid3s3wf5m98hcpbvgs8mk-clang-wrapper-16.0.6/bin/cc -Wall -Werror -Wno-tautological-compare -ObjC -o /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c' output included error: /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c:34:10: fatal error: 'Foundation/Foundation.h' file not found
Error:      C file contents around line 34:
Error:      /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c:33: #include <unistd.h>
Error:      /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c:34: #include <Foundation/Foundation.h>
Error:      /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c:35: #include <mach/mach.h>
Error:  com.oracle.svm.core.util.UserError$UserException: Error compiling query code (in /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c). Compiler command '/nix/store/c1ib07kj7baid3s3wf5m98hcpbvgs8mk-clang-wrapper-16.0.6/bin/cc -Wall -Werror -Wno-tautological-compare -ObjC -o /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c' output included error: /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c:34:10: fatal error: 'Foundation/Foundation.h' file not found
Error:      C file contents around line 34:
Error:      /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c:33: #include <unistd.h>
Error:      /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c:34: #include <Foundation/Foundation.h>
Error:      /var/folders/q5/6nls1_g145s2fl1lbvx2zby40000gn/T/SVM-14881555311105086393/PosixDirectives.c:35: #include <mach/mach.h>
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:138)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.NativeLibraries.reportErrors(NativeLibraries.java:380)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.processNativeLibraryImports(NativeImageGenerator.java:1791)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.setupNativeLibraries(NativeImageGenerator.java:1237)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:979)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:569)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:537)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:526)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:701)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.start(NativeImageGeneratorRunner.java:140)
Error:  	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:95)
```

--- a/pkgs/build-support/build-graalvm-native-image/default.nix
+++ b/pkgs/build-support/build-graalvm-native-image/default.nix
@@ -53,7 +53,7 @@ stdenv.mkDerivation ({
   buildPhase = args.buildPhase or ''
     runHook preBuild
 
-    native-image -jar "$jar" $(export -p | sed -n 's/^declare -x \([^=]\+\)=.*$/ -E\1/p' | tr -d \\n) ''${nativeImageBuildArgs[@]}
+    native-image -jar "$jar" ''${nativeImageBuildArgs[@]}
 
     runHook postBuild
   '';

--- a/pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix
+++ b/pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix
@@ -120,27 +120,44 @@ let
       xorg.libXtst
     ];
 
-    postInstall = ''
-      # jni.h expects jni_md.h to be in the header search path.
-      ln -sf $out/include/linux/*_md.h $out/include/
+    postInstall =
+      let
+        cLibsAsFlags = (map (l: "--add-flags '-H:CLibraryPath=${l}/lib'") cLibs);
+        preservedNixVariables = [
+          "-ELOCALE_ARCHIVE"
+          "-ENIX_BINTOOLS"
+          "-ENIX_BINTOOLS_WRAPPER_TARGET_HOST_${stdenv.cc.suffixSalt}"
+          "-ENIX_BUILD_CORES"
+          "-ENIX_BUILD_TOP"
+          "-ENIX_CC"
+          "-ENIX_CC_WRAPPER_TARGET_HOST_${stdenv.cc.suffixSalt}"
+          "-ENIX_CFLAGS_COMPILE"
+          "-ENIX_HARDENING_ENABLE"
+          "-ENIX_LDFLAGS"
+        ];
+        preservedNixVariablesAsFlags = (map (f: "--add-flags '${f}'") preservedNixVariables);
+      in
+      ''
+        # jni.h expects jni_md.h to be in the header search path.
+        ln -sf $out/include/linux/*_md.h $out/include/
 
-      mkdir -p $out/share
-      # move files in $out like LICENSE.txt
-      find $out/ -maxdepth 1 -type f -exec mv {} $out/share \;
-      # symbolic link to $out/lib/svm/LICENSE_NATIVEIMAGE.txt
-      rm -f $out/LICENSE_NATIVEIMAGE.txt
+        mkdir -p $out/share
+        # move files in $out like LICENSE.txt
+        find $out/ -maxdepth 1 -type f -exec mv {} $out/share \;
+        # symbolic link to $out/lib/svm/LICENSE_NATIVEIMAGE.txt
+        rm -f $out/LICENSE_NATIVEIMAGE.txt
 
-      # copy-paste openjdk's preFixup
-      # Set JAVA_HOME automatically.
-      mkdir -p $out/nix-support
-      cat > $out/nix-support/setup-hook << EOF
-      if [ -z "\''${JAVA_HOME-}" ]; then export JAVA_HOME=$out; fi
-      EOF
+        # copy-paste openjdk's preFixup
+        # Set JAVA_HOME automatically.
+        mkdir -p $out/nix-support
+        cat > $out/nix-support/setup-hook << EOF
+        if [ -z "\''${JAVA_HOME-}" ]; then export JAVA_HOME=$out; fi
+        EOF
 
-      wrapProgram $out/bin/native-image \
-        --prefix PATH : ${binPath} \
-        ${toString (map (l: "--add-flags '-H:CLibraryPath=${l}/lib'") cLibs)}
-    '';
+        wrapProgram $out/bin/native-image \
+          --prefix PATH : ${binPath} \
+          ${toString (cLibsAsFlags ++ preservedNixVariablesAsFlags)}
+      '';
 
     preFixup = lib.optionalString (stdenv.hostPlatform.isLinux) ''
       for bin in $(find "$out/bin" -executable -type f); do
@@ -171,10 +188,8 @@ let
       echo "Testing GraalVM"
       $out/bin/java -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -XX:+UseJVMCICompiler HelloWorld | fgrep 'Hello World'
 
-      extraNativeImageArgs="$(export -p | sed -n 's/^declare -x \([^=]\+\)=.*$/ -E\1/p' | tr -d \\n)"
-
       echo "Ahead-Of-Time compilation"
-      $out/bin/native-image -H:+UnlockExperimentalVMOptions -H:-CheckToolchain -H:+ReportExceptionStackTraces -march=compatibility $extraNativeImageArgs HelloWorld
+      $out/bin/native-image -H:+UnlockExperimentalVMOptions -H:-CheckToolchain -H:+ReportExceptionStackTraces -march=compatibility HelloWorld
       ./helloworld | fgrep 'Hello World'
 
       ${# -H:+StaticExecutableWithDynamicLibC is only available in Linux

